hpx/parallel/util/foreach_partitioner.hpp

```diff
@@ -38,31 +38,33 @@ namespace hpx { namespace parallel { namespace util
             static FwdIter call(ExPolicy policy, FwdIter first,
                 std::size_t count, F1 && f1, std::size_t chunk_size)
             {
+                typedef typename ExPolicy::executor_type executor_type;
+                typedef typename hpx::parallel::executor_traits<executor_type>
+                    executor_traits;
+
+                FwdIter last = first;
+                std::advance(last, count);
+
                 std::vector<hpx::future<Result> > workitems;
                 std::list<boost::exception_ptr> errors;
 
                 try {
-                    // estimate a chunk size based on number of cores used
-                    chunk_size = get_static_chunk_size(policy, workitems, f1,
+                    std::vector<std::pair<FwdIter, std::size_t> > shape;
+                    std::vector<hpx::future<Result> > inititems;
+                    // estimates a chunk size base on number of cores used
+                    shape = get_static_shape(policy, inititems, f1,
                         first, count, chunk_size);
 
-                    // schedule every chunk on a separate thread
-                    workitems.reserve(count / chunk_size + 1);
-                    while (count != 0)
+                    auto f = [f1](std::pair<FwdIter, std::size_t> elem)
                     {
-                        std::size_t chunk = (std::min)(chunk_size, count);
-
-                        typedef typename ExPolicy::executor_type executor_type;
-                        workitems.push_back(
-                            executor_traits<executor_type>::async_execute(
-                                policy.executor(),
-                                hpx::util::deferred_call(f1, first, chunk)
-                            )
-                        );
-
-                        count -= chunk;
-                        std::advance(first, chunk);
-                    }
+                        return f1(elem.first, elem.second);
+                    };
+
+                    workitems = executor_traits::async_execute(
+                        policy.executor(), f, shape);
+
+                    std::move(inititems.begin(), inititems.end(),
+                        std::back_inserter(workitems));
```

Comment:
While having `workitems` and `inititmes` isn't a problem for the `foreach_partition`, what is the best way to handle that when calling `f2` for `partitioner`? The logic for many of the algorithms requires one vector with all the results from `f1` to call `f2` .